src/modules/compliance/src/lib/procedures/packageInstalled.cpp

```diff
@@ -3,31 +3,357 @@
+#include <linux/limits.h>
+#include <set>
 #include <string>
+#include <unistd.h>
+
+namespace
+{
+
+using compliance::Error;
+using compliance::Result;
+
+typedef struct
+{
+    time_t lastUpdateTime;
+    std::string packageManager;
+    std::set<std::string> packageNames;
+} PackageCache;
+
+static constexpr long PACKAGELIST_TTL = 3000L;        // just shy of an hours
+static constexpr long PACKAGELIST_STALE_TTL = 12600L; // 3.5 hours
+static constexpr char PACKAGE_CACHE_PATH[] = "/var/lib/osconfig/packageCache";
+
+std::string DetectPackageManager(OsConfigLogHandle log)
+{
+    if (ExecuteCommand(NULL, "dpkg -l dpkg", false, false, 0, 0, NULL, NULL, log) == 0)
+    {
+        return "dpkg";
+    }
+    else if (ExecuteCommand(NULL, "rpm -qa rpm", false, false, 0, 0, NULL, NULL, log) == 0)
+    {
+        return "rpm";
+    }
+    return "";
+}
+
+Result<PackageCache> LoadPackageCache(const std::string& path)
+{
+    PackageCache cache;
+    const std::string pkgCacheHeader = "# PackageCache "; // "# PackageCache <packageManager>@<timestamp>\n"
+    cache.lastUpdateTime = 0;
+    cache.packageManager = "";
+    cache.packageNames.clear();
+    std::ifstream cacheFile(path);
+    if (!cacheFile.is_open())
+    {
+        return Error("Failed to open cache file: " + path);
+    }
+
+    std::string header;
+    if (!std::getline(cacheFile, header) || (0 != header.rfind(pkgCacheHeader, 0)))
+    {
+        return Error("Invalid cache file format");
+    }
+
+    auto separatorPos = header.find('@');
+    if (std::string::npos == separatorPos)
+    {
+        return Error("Invalid cache file header format");
+    }
+
+    cache.packageManager = header.substr(pkgCacheHeader.length(), separatorPos - pkgCacheHeader.length());
+    try
+    {
+        cache.lastUpdateTime = std::stol(header.substr(separatorPos + 1));
+    }
+    catch (const std::exception&)
+    {
+        return Error("Invalid timestamp in cache file header");
+    }
+
+    std::string packageName;
+    while (std::getline(cacheFile, packageName))
+    {
+        if (!packageName.empty())
+        {
+            cache.packageNames.insert(packageName);
+        }
+    }
+
+    if (cacheFile.bad())
+    {
+        return Error("Error reading cache file");
+    }
+
+    return cache;
+}
+
+Result<int> SavePackageCache(const PackageCache& cache, const std::string& path, OsConfigLogHandle log)
+{
+    std::string templatePath = path + ".XXXXXX";
+    char pathTemplate[PATH_MAX];
+    snprintf(pathTemplate, sizeof(pathTemplate), "%s", templatePath.c_str());
+    int fd = mkstemp(pathTemplate);
+    if (fd == -1)
+    {
+        return Error("Failed to create temporary file");
+    }
+    close(fd);
+
+    templatePath = pathTemplate;
+    std::string tempPath = templatePath + ".tmp";
+    std::ofstream tempFile(tempPath, std::ios::out | std::ios::trunc);
+    if (!tempFile.is_open())
+    {
+        return Error("Failed to open temporary file for writing: " + tempPath);
+    }
```

Comment:
This usage of mkstemp seems suspicious because you end up opening a different file (so you're racing), and you never cleanup the file created by `mkstemp` (it doesn't get cleaned up automagically).
How about something more like this:
```
int fd = mkstemp(...)
if (fd == -1)
{
...
}
std::ofstream tempFile(pathTemplate, ...);
close(fd);
```

Reply:
Done.